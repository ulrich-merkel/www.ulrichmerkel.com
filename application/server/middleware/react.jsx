/**
 * Es6 module for react server middleware.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.3
 *
 * @requires react
 * @requires react-dom
 * @requires react-router
 * @requires lodash
 * @requires common/config/application
 * @requires common/config/routes
 * @requires common/utils/logger
 * @requires common/utils/read-file
 * @requires common/component/root
 * @requires common/component/layout/html
 * @requires common/component/page/not-found
 * @requires common/state/configure-store
 * @requires common/state/config/actions
 * @requires common/state/intl/actions
 * @requires common/state/csrf/actions
 *
 * @see {@link https://github.com/reactjs/react-router/blob/1.0.x/docs/guides/advanced/ServerRendering.md}
 * @see {@link https://github.com/reactjs/react-router/issues/1990}
 * @see {@link https://www.youtube.com/watch?v=PnpfGy7q96U}
 * @see {@link https://github.com/reactjs/redux/issues/723}
 *
 * @changelog
 * - 0.0.3 Adjusted async rendering
 * - 0.0.2 Moved code to es6
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { RouterContext, match } from 'react-router';
import { get } from 'lodash';

import configApplication from './../../common/config/application';
import configRoutes from './../../common/config/routes';
import logger from './../../common/utils/logger';
import { readFile } from './../../common/utils/read-file';
import Root from './../../common/component/root';
import LayoutHtml from './../../common/component/layout/html';
import PageNotFound from './../../common/component/page/not-found';
import configureStore from './../../common/state/configure-store';
import { changeLocale } from './../../common/state/intl/actions';
import { fetchConfigContentIfNeeded, fetchConfigTranslationIfNeeded } from './../../common/state/config/actions';
import { addToken } from './../../common/state/csrf/actions';

const { aboveTheFold } = configApplication;

/**
 * Helper function to pass props to children components.
 *
 * @function
 * @private
 * @param {Object} Component The current router react component
 * @param {Object} props The current react component props
 * @returns {React.Element} React component markup
 */
function createElement(Component, props) {
    return <Component {...props} />;
}

/**
 * Using createElement function here to be able to
 * pass props to the page Component
 *
 * @see {@link https://github.com/reactjs/react-router/issues/3183}
 * @see {@link https://github.com/voronianski/universal-react-router-flux-2016/blob/master/src/server/middleware/renderHTML.js#L17}
 * @see {@link https://github.com/ReactTraining/react-router/issues/902}
 *
 * @function
 * @private
 * @param {Object} store The created redux store
 * @param {Object} renderProps The component properties to be rendered
 * @param {string} [cssBase=''] The file contents from base.css
 * @param {string} [scriptLoader=''] The file contents from loader.js
 * @returns {void}
 */
function getHtml(store, renderProps, cssBase = '', scriptLoader = '') {
    return renderToStaticMarkup(
        <Root store={store}>
            <LayoutHtml {... { store, cssBase, scriptLoader }}>
                <RouterContext
                    {...renderProps}
                    createElement={createElement}
                />
            </LayoutHtml>
        </Root>
    );
}

/**
 * Handle react server rendering and react routering.
 *
 * @function
 * @param {Object} req The current request object
 * @param {Object} res The result object
 * @returns {void}
 */
function middlewareReact(req, res, next) {

    /**
     * Using the new match function instead of Router.run
     *
     * Note that req.url here should be the full URL path from
     * the original request, including the query string.
     *
     * @see {@link https://github.com/reactjs/react-router/blob/1.0.x/docs/guides/advanced/ServerRendering.md}
     * @see {@link https://github.com/reactjs/react-router/blob/master/docs/guides/ServerRendering.md}
     */
    match({
        routes: configRoutes,
        location: req.url
    }, (error, redirectLocation, renderProps) => {

        if (error) {
            return res.status(error.status || 500).send(error.message).end();
        }

        if (redirectLocation) {
            return res.redirect(res.status || 302, `${redirectLocation.pathname}${redirectLocation.search}`);
        }

        if (renderProps) {

            // Initialize redux store, combining single stores into a global one
            const store = configureStore();

            /**
             * Read language which is given in the url locale parameter or via
             * express middleware and store the selection in redux
             */
            const urlLocale = get(renderProps, 'router.params.locale', '');
            store.dispatch(changeLocale([urlLocale, urlLocale.toUpperCase()].join('-'), req.language));
            const acceptedLocale = get(store.getState(), 'intl.locale');

            /**
             * We check renderProps.components (or renderProps.routes) for
             * the "not found" component or route respectively, and send a 404 as
             * below, because we're using a catch-all error route in the router
             * config.
             *
             * @see {@link https://github.com/ReactTraining/react-router/issues/2834}
             */
            const statusCode = get(renderProps, 'components', []).indexOf(PageNotFound) !== -1 ? 404 : 200;

            /**
             * Load all required data async via promises to
             * improve the overall node performance.
             */
            return Promise.all([
                readFile(aboveTheFold.baseCss),
                readFile(aboveTheFold.loaderBundle),
                store.dispatch(fetchConfigContentIfNeeded()),
                store.dispatch(fetchConfigTranslationIfNeeded(acceptedLocale)),
                store.dispatch(addToken(req.csrfToken && req.csrfToken()))
            ]).then((result) => {
                return res
                    .status(statusCode)
                    .send(`<!doctype html>${getHtml(store, renderProps, result[0], result[1])}`);
            }).catch((reason) => {
                logger.warn(reason);
                return res
                    .status(statusCode)
                    .send(`<!doctype html>${getHtml(store, renderProps)}`);
            });

        }

        return next();

    });
}

export default middlewareReact;
