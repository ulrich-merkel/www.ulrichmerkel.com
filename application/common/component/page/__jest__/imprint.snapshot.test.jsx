/* eslint-disable import/no-extraneous-dependencies, func-names */
import * as React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import mockedStore from '../../__mocks__/store';
import PageImprint from '../imprint';

describe('common/component/page/imprint', function () {
    it('should render correctly', function () {
        const tree = renderer
            .create(
                <Provider store={mockedStore}>
                    <MemoryRouter>
                        <PageImprint />
                    </MemoryRouter>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
