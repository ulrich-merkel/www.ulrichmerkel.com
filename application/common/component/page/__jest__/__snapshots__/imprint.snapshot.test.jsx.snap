// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/component/page/imprint should render correctly 1`] = `
<main
  className="l-main"
  itemProp="mainContentOfPage"
  role="main"
>
  <div
    className="l-main__transition"
  >
    <section
      className="l-grid m-section"
    >
      <div
        className="l-grid__spaced--small"
      >
        <div
          className="l-grid__row"
        >
          <div
            className="l-grid__col--12"
          >
            <article
              className="is-main m-article"
              itemScope={true}
              itemType="https://schema.org/Article"
            >
              <h1
                className="is-centered c-type--h1 m-article__headline"
                itemProp="headline"
              >
                Imprint
              </h1>
              <p
                className="is-centered c-type--p m-article__lead"
                itemProp="alternativeHeadline"
              >
                <strong>
                  Legal details and information in accordance with section 5 TMG.
                </strong>
              </p>
              <div
                className="m-article__text"
                itemProp="text"
              >
                <div
                  className="m-text"
                >
                  <div
                    className="m-text__block"
                  >
                    <h3
                      className="is-centered c-type--h3 m-text__headline"
                    >
                      Publisher information
                    </h3>
                    <div
                      className="is-centered m-text__content"
                    >
                      <p
                        className="c-type--p"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "Person responsible for content in accordance with 55 Abs. 2 RStV:",
                          }
                        }
                      />
                    </div>
                    <div
                      className="is-centered m-person is-centered m-text__content"
                      itemScope={true}
                      itemType="http://schema.org/Person"
                    >
                      <p
                        className="c-type--p m-person__name"
                      >
                        <strong>
                          Ulrich Merkel
                        </strong>
                      </p>
                      <address
                        className="m-person__address c-type--address"
                        itemProp="address"
                        itemScope={true}
                        itemType="http://schema.org/Address"
                      >
                        <span
                          className="m-person__street-address"
                          itemProp="street-address"
                        >
                          Pegauer Straße 51
                        </span>
                        <span
                          className="m-person__postal-code"
                          itemProp="postal-code"
                        >
                          04442
                        </span>
                        <span
                          className="m-person__locality"
                          itemProp="locality"
                        >
                          Zwenkau, Germany
                        </span>
                      </address>
                      <p
                        className="c-type--p m-person__email"
                      >
                        <abbr
                          title="E-Mail address"
                        >
                          E.
                        </abbr>
                         
                        <a
                          href="mailto:hello@ulrichmerkel.com"
                          itemProp="email"
                        >
                          hello@ulrichmerkel.com
                        </a>
                      </p>
                      <p
                        className="c-type--p m-person__website"
                      >
                        <abbr
                          title="Website"
                        >
                          W.
                        </abbr>
                         
                        <a
                          href="http://www.ulrichmerkel.com"
                        >
                          http://www.ulrichmerkel.com
                        </a>
                      </p>
                    </div>
                    <div
                      className="is-centered m-text__content"
                    >
                      <p
                        className="c-type--p"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "Ust. (VAT) - Idnr. DE  235 248 02155<br />
        Amtsgericht Borna",
                          }
                        }
                      />
                    </div>
                  </div>
                  <div
                    className="m-text__block"
                  >
                    <h3
                      className="is-centered c-type--h3 m-text__headline"
                    >
                      Indication of source for images and graphics
                    </h3>
                    <div
                      className="is-centered m-text__content"
                    >
                      <p
                        className="c-type--p"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "<a href='http://www.adrian-kuehne.de/' target='_blank' title='Show Adrian Kühne website' rel='noopener noreferrer'>
        <i aria-hidden='true' class='c-font-icon--earth'></i> http://www.adrian-kuehne.de/ (Adrian Kühne)
        </a>",
                          }
                        }
                      />
                    </div>
                  </div>
                  <div
                    className="m-text__block"
                  >
                    <h3
                      className="is-centered c-type--h3 m-text__headline"
                    >
                      Technology
                    </h3>
                    <div
                      className="is-centered m-text__content"
                    >
                      <p
                        className="c-type--p"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "This website is build with React.js, Node.js, <abbr title='HyperText Markup Language 5'>HTML5</abbr>, <abbr title='Cascading Style Sheets Level 3'>CSS3</abbr>, love and 100% recycled pixels.
        View <a href='https://github.com/ulrich-merkel/www.ulrichmerkel.com' target='_blank' title='Show source code on GitHub' rel='noopener noreferrer'>source code</a> on Github.",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
              </div>
              <p
                className="is-centered c-type--p m-article__button hide-on-print"
              >
                <a
                  aria-current="true"
                  className="c-btn active"
                  href="/"
                  onClick={[Function]}
                  role="button"
                  style={Object {}}
                  title="Go to startpage"
                  type={null}
                >
                  <span
                    className="c-btn__label"
                  >
                    Back to startpage
                  </span>
                </a>
              </p>
              <meta
                content="Imprint"
                itemProp="name"
              />
              <meta
                content="2017-10-7"
                itemProp="datePublished"
              />
            </article>
          </div>
        </div>
      </div>
    </section>
  </div>
</main>
`;
